content/blog/en/dojo-middleware.md

```diff
@@ -0,0 +1,213 @@
+
+import PhoneValidator from './PhoneValidator';
+
+import * as css from './styles/PhoneNumber.m.css';
+
+const factory = create({ icache });
+
+export const PhoneNumber = factory(function PhoneNumber({ middleware: { icache } }) {
+	// use local phone value to pass to validator
+	const phone = icache.getOrSet('phone', '');
+	return (
+		<PhoneValidator phone={phone}>
+			<input
+				placeholder="Enter phone number"
+				pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"
+				required
+				classes={[css.root]}
+				type="tel"
+				onkeyup={(e) => {
+					icache.set('phone', (e.target as HTMLInputElement).value);
+				}}
+			/>
+		</PhoneValidator>
+	);
+});
+
+export default PhoneNumber;
+```
+
+The idea here is that we want the `PhoneValidator` to place a red or green outline to my input to let me know if it's a valid phone number or not. It's pretty simple, but is something that could be reused across multiple applications.
+
+> src/widgets/PhoneValidator.tsx
+
+```tsx
+import { create, tsx } from '@dojo/framework/core/vdom';
+import phoneNumberMiddleware from '../middleware/phoneNumberMiddleware';
+
+import * as css from './styles/PhoneValidator.m.css';
+
+interface Properties {
+	phone: string;
+}
+
+const factory = create({ phoneNumberMiddleware }).properties<Properties>();
+
+export const PhoneValidator = factory(function PhoneValidator({ children, middleware: { phoneNumberMiddleware } }) {
+	const { valid, value } = phoneNumberMiddleware();
+	let validCss = '';
+	if (value.length) {
+		validCss = valid ? css.valid : css.invalid;
+	}
+	return <div classes={[css.root, validCss]}>{children()}</div>;
+});
+
+export default PhoneValidator;
+```
+
+The `PhoneValidator` uses some middleware that returns the a `valid` property that is either `true` or `false`. It will also return the `value` of the phone number that was tested. Based on whether the phone number is valid or not, it will use some CSS for a red or green border.
+
+Notice that the `phone` property to the middleware is not passed explicitly to the middleware. By provide the `phoneNumberMiddleware` as a middleware to the `PhoneValidator` widget, the middleware will have access to the properties of the widget. Let's see what that looks like.
+
+> src/middleware/phoneNumberMiddleware.tsx
+
+```tsx
+import { create } from '@dojo/framework/core/vdom';
+
+const factory = create().properties<{ phone?: string }>();
+
+export const phoneNumberMiddleware = factory(({ properties }) => {
+	return () => {
+		// extract the `phone` value from the properties of
+		// the parent widget
+		const { phone } = properties();
+		// test the phone number
+		const valid = /^\(?(\d{3})\)?[- ]?(\d{3})[- ]?(\d{4})$/.test(phone || '');
+		return {
+			valid,
+			value: phone
+		};
+	};
+});
+
+export default phoneNumberMiddleware;
+```
+
+The middleware returns a function that will test the phone number and return whether it is valid or not.
+
+Here is what this looks like in a sample application.
```

Comment:
```suggestion
View this middleware example in a sample phone validation application.
```

Reply:
We might also want to link "sample phone validation application" to an example page on codesandbox in case they're on a small screen?